Создание слоя доступа к данным.

sFedyashov opened the pull request:
Создан слой работы с данными. В самом приложении/тестах работа только
через него. Операции, специфичные для конкретной СУБД (MongoDB) убраны в
отдельный датапровайдер. Тем самым убираем связность (Issue #29) и
готовим приложение для работы с другими БД (Issue #28)

sFedyashov committed:
Создание слоя доступа к данным.

Создан слой работы с данными. В самом приложении/тестах работа только
через него. Операции, специфичные для конкретной СУБД (MongoDB) убраны в
отдельный датапровайдер. Тем самым убираем связность (Issue #29) и
готовим приложение для работы с другими БД (Issue #28)

ExM commented on .gitignore:
```diff
@@ -11,3 +11,7 @@ TestResults
+*.bin
+*.dat
+*.sst
+*.version
```
Зачем добавлены эти исключения? Какие именно инструменты разработчика приводят к появлению таких файлов?

sFedyashov replied:
решарпер. в связи с этим и замечанием ниже почитаю, исправлю.

ExM commented on BuildRevisionCounter.Tests/Controllers/CounterControllerTest.cs:
```diff
@@ -14,9 +14,9 @@ public class CounterControllerTest
-			MongoDBStorageUtils.SetUpAsync().Wait();
-
-			_controller = new CounterController(MongoDBStorageFactory.DefaultInstance);
+            DBStorageFactory.DefaultInstance.SetUpAsync().Wait();
```
Нарушение CodeStyle.
Надо внимательно прочитать https://github.com/ExM/BuildRevisionCounter/blob/master/CONTRIBUTING.md
а также документ по ссылке в нем http://rsdn.ru/article/mag/200401/codestyle.XML

ExM commented on BuildRevisionCounter/Interfaces/IDataProvider.cs:
```diff
@@ -0,0 +1,36 @@
+	    Task CreateUser(string name, string password, string[] roles);
+
+	    Task EnsureAdminUser();
+    }
```
Интерфейс отвечает за разнородный функционал.
Какие здесь есть варианты разделения функционала?

ExM commented:
файл _ReSharper.BuildRevisionCounter/PersistentCaches/MANIFEST-000013
не следует коммитить в репо. Если пользуешся решарпером, то поищи типовые паттерны для .gitignore

ExM commented on BuildRevisionCounter/Data/DbStorage.cs:
```diff
@@ -0,0 +1,110 @@
+
+namespace BuildRevisionCounter.Data
+{
+    public class DbStorage
```
В части всех вызовов - это класс обертка над интерфейсом IDataProvider, за исключением метода Bumping, который мог быть и в составе интерфейса.
Почему нельзя было использовать сам интерфейс IDataProvider?

sFedyashov replied:
мысль была такая:
IDataProvider - это классы, которые знают как работать с конкретным хранилищем, реализуют какие-то низкоуровневые операции. их может быть несколько.
DBStorage - обёртка, которая по строке из конфига (сейчас зашито в коде) определяет, какой именно класс использовать, из низкоуровневых операций делает более крупные, содержит какую-то дополнительную логику.

ExM replied:
Для укрупнения операций здесь надо использовать extension к интерфейсу. Здесь нет инкапсуляции.

ExM replied:
Bumping будет сильно зависить от реализации хранилища. В SQL это может быть один запрос, но мы интерфейсно принуждаем его разбивать на 2.

ExM commented on BuildRevisionCounter/Interfaces/IDataProvider.cs:
```diff
@@ -0,0 +1,36 @@
+
+	    Task SetUp();
+
+	    Task DropDatabaseAsync();
```
Такого функционала не должно быть в боевом коде.

sFedyashov committed:
функционал IDataProvider'а разделил на два интерфейса, убрал класс-прослойку, исправил своё ошибочное форматиривание

- разделил функционал на два блока: "рабочий" и "тестовый" (теперь
отдельные интерфейсы)
- убрал класс-прослойку, часть его функционала перенёс в расширяемые
методы (extension)
- в проекте тестирования создание DataProvider стало по классу - теперь
возможно тестировать сразу несколько классов/несколько субд
- поправил своё прошлое ошибочное форматирование (студия заменяла табы
на пробелы - теперь настроил её правильно)
- в gitignore файлы решарпера убрал с помощью "_ReSharper*"

ExM commented on BuildRevisionCounter.Tests/DBStorageFactory.cs:
```diff
@@ -0,0 +1,42 @@
+		}
+
+		public static IDatabaseTestProvider FromConfigurationConnectionString(string connectionStringName = "MongoDBStorage")
+        {
```
пробелы еще остались

ExM commented on BuildRevisionCounter/Interfaces/IDataProvider.cs:
```diff
@@ -0,0 +1,23 @@
+	/// <summary>
+	/// Интерфейс для получения данных из БД.
+	/// </summary>
+	public interface IDataProvider
```
Интерфейс все еще отвечает за разнородный функционал.
А в здесь можно обсудить какие же есть варианты разделения функционала.
Например:
- операции с ревизиями
- операции с пользователями

sFedyashov replied:
я считал, что раз там 5 методов и больше 10 не ожидается, то это однородный функционал "работа с БД".
разделил в "боевом" интерфейсе, в тестовом не трогал, посчитав, что это "функционал БД, который надо протестировать".

ExM commented on BuildRevisionCounter/Interfaces/IDatabaseTestProvider.cs:
```diff
@@ -0,0 +1,23 @@
+	/// <summary>
+	/// Интерфейс для проверки операций с БД.
+	/// </summary>
+	public interface IDatabaseTestProvider
```
Кажется, что EnsureAdminUser может быть реализован в виде расширения.
CreateUser пригодится в IDataProvider, все таки планируется разработать UI управления пользователями (#27)

ExM commented on BuildRevisionCounter/Interfaces/IDatabaseTestProvider.cs:
```diff
@@ -0,0 +1,23 @@
+
+			Task SetUp();
+
+			Task DropDatabaseAsync();
```
Не следует оставлять такой функционал в интерфейсе который определен в "боевой" сборке.
Это метод можно просто вызвать у конкретной реализации IDataProvider в коде тестовой сборки.

Можно ли вообще избавиться от этого метода в коде?

sFedyashov replied:
мне кажется, что нас не спасёт, если эти методы всё так же останутся публичными и просто перестанут относиться к какому-то интерфейсу.
В голове крутятся варианты:
-сделать метод приватным и вызывать его через рефлекшн (как-то не очень хорошо)
-вообще перенести этот код в тестовую сборку (не хочется работу с БД разделять)

В итоге интерфейс оставил, а возможное ошибочное обращение к методу закрыл через атрибут:
[MethodUsageScope(MainModuleName = "nunit-agent.exe")]
public Task DropDatabaseAsync()
{
MethodUsageScopeAttribute.ControlScope(MethodBase.GetCurrentMethod());

в случае, если тестирование будет не через модуль "nunit-agent.exe", то возможно потребуется правка (предусмотрел перечисление имён через запятую).

ExM replied:
Этот подход еще хуже, т.к. намертво привязывает боевой код к определенным тестам.

Итак, задача: избавиться от наличия в боевом коде метода удаления БД.

Чистая БД требуется для запуска тестов, в которых зафиксированны имена тестируемых сущностей.
К этим именам надо прибавить случайную составляющую. (например GUID)
Для тестирования генерации первого пользователя следует реализовать методы удаления пользователя и получения списка пользователей.

sFedyashov replied:
тестирование всех методов сделал на новой БД: первый пользователь теперь создаётся в новой пустой БД, а сама БД удаляется после завершения операции тестирования (т.к. строкой подключения мы можем легко оперировать, а названия сущностей более глубоко зашиты и местами до них совсем тяжело добраться: например в IntegrationTest).

ExM replied:
Сформулируем иначе.

ExM replied:
Сформулируем иначе.
Следует вообще избавиться от удаления БД. При анализе прохождения тестов в отчете билд сервера, очень полезно иметь в наличии незачищенную БД.

названия (имя пользователя, имя счетчика) в каждом тесте следует расширить случайной составляющей. (например GUID)

sFedyashov replied:
после теста эти строки в таблицах удалять или оставлять?

ExM replied:
Т.к. они не мешают, или не должны при правильном коде, то можно оставить.
Побочный эффект - контент для разработки UI.

ExM commented on BuildRevisionCounter/Data/MongoDBStorage.cs:
```diff
@@ -0,0 +1,155 @@
+			return revision.CurrentNumber;
+		}
+
+		public async Task RevisionInsertAsync(string revisionName)
```
Метод может вернуть результат операции в виде bool, это упростит его вызов.

sFedyashov replied:
В виде похожем на int.TryParse(string, out int) сделать не можем, т.к. метод асинхронный и аут параметры не использовать. Можно сделать класс/структуру (или использовать Tuple) для возвращаемого значения, но тогда мы получим тоже самое, что уже имеем (аналогичные конструкции проверки и присваивания результата).

ExM replied:
Вообще то, я комментировал "RevisionInsertAsync". ;)
К "CurrentRevision" претензхий не было.

ExM commented on BuildRevisionCounter/Data/DbProviderUtil.cs:
```diff
@@ -0,0 +1,51 @@
+			return (IDataProvider)Activator.CreateInstance(type, connectionString);
+		}
+		
+		public static async Task<long> Bumping(this IDataProvider dataProvider, string revisionName)
```
Все таки почему Bumping остался реализован расширением?

sFedyashov committed:
разделил "боевой" интерфейс, разграничил доступ к методам тестирования БД, рефакторинг

- разделил "боевой" интерфейс на работу с ревизиями и с пользователями
- разграничил доступ к методам тестирования БД (через атрибут с
перечислением имён разрешенных для работу прорамм)
- рефакторинг (пробелы, EnsureAdminUser, Bumping)

ExM commented on BuildRevisionCounter/Data/MongoDBStorage.cs:
```diff
@@ -0,0 +1,170 @@
+
+namespace BuildRevisionCounter.Data
+{
+	public class MongoDBStorage : IRevisionDataProvider, IUserDataProvider, IDatabaseTestProvider
```
Класс отвечает за разнородный функционал.
Букву I в SOLID мы уже реализовали, теперь обратим внимание на букву S

ExM commented on BuildRevisionCounter/Interfaces/IUserDataProvider.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace BuildRevisionCounter.Interfaces
+{
+	public interface IUserDataProvider
```
Слишком общее название. Технически это просто хранилище объектов одного класса.
Можно назвать IUserRepository, IUserStorage, IUserCollection

ExM commented on BuildRevisionCounter.Tests/DBStorageFactory.cs:
```diff
@@ -0,0 +1,42 @@
+			if (type == null)
+				throw new ApplicationException("на найден класс для IDataProvider");
+			var connectionString = ConfigurationManager.ConnectionStrings[connectionStringName].ConnectionString;
+			return (IDatabaseTestProvider)Activator.CreateInstance(type, connectionString);
```
Зачем использовать Activator, если в этом месте есть доступ до требуемого типа?

sFedyashov replied:
мысль аналогична следующему (про DbProviderUtil.cs).

ExM commented on BuildRevisionCounter/Data/DbProviderUtil.cs:
```diff
@@ -0,0 +1,54 @@
+			if (type == null)
+				throw new ApplicationException("на найден класс для IDataProvider");
+			var connectionString = ConfigurationManager.ConnectionStrings[connectionStringName].ConnectionString;
+			return (IRevisionDataProvider)Activator.CreateInstance(type, connectionString);
```
Зачем использовать Activator, если в этом месте есть доступ до требуемого типа?
Очень сильно ухудшаюется возможность рефакторинга, т.к. ошибки при изменении имени класса или состава параметров конструктора переносятся в runtime.

sFedyashov replied:
По задумке в дальнейшем может быть несколько провайдеров/репозиториев, которые могут работать с разными БД. Какой именно провайдер использовать для работы с БД будет определяться по строке с именем класса из конфига. Про эту "подготовку к работе с несколькими БД" я писал в первом коммите (для дальнейших работ по Issue #28). Для лучшего понимания этого подхода другими разработчиками (которые будут делать Issue #28) могу сделать эту настройку в конфиге и читать значение оттуда.
Если этот подход не нравится, то могу:
- зашить в коде конкретный класс, выбор БД будет осуществляться правкой кода перед компиляцией (пока класс/БД и так только один)
- придумать что-то ещё

ExM replied:
Сейчас следует зашить в коде конкретный класс.
В дальнейшем (не в этом issue), можно будет разработать методику динамической подгрузка сборки с требуемой реализацией.

sFedyashov committed:
правки по замечаниям

следующие правки:
-разделение класса работы с БД на 2 (ревизии и пользователи)
-код удаления БД перенесён в тестирующую сборку
-для тестов создаётся БД со случайным именем (..._GUID), после тестов
удаляется.
-переименования интерфейсов

ExM commented on BuildRevisionCounter/Data/DBStorageUtil.cs:
```diff
@@ -0,0 +1,39 @@
+
+		public static string GetAdminName(this IUserDatabaseTestProvider dataProvider)
+		{
+			return "admin";
```
Как результат метода зависит от его аргумента?

sFedyashov replied:
аргумент для привязывания расширения к интерфейсу.

ExM replied:
А зачем его привязывать? Метод не возможно перегрузить при наследовании или реализовать при реализации методов интерфейса, почему просто не воспользоваться константой?

ExM commented on BuildRevisionCounter/Interfaces/IUserDatabaseTestProvider.cs:
```diff
@@ -0,0 +1,19 @@
+			Task<UserModel> FindUser(string name);
+			
+			Task CreateUser(string name, string password, string[] roles);
+	}
```
последние два метода присутствуют в IUserStorage, зачем они здесь?

sFedyashov replied:
для того, чтобы работать с одним интерфейсом тестирования БД

ExM replied:
Интерфейсы можно наследовать. 
Недостатки этого кода:
- еще один интерфейс для настройки, который не используется в боевом коде.
- в боевом коде используется не тот интерфейс который тестируется.

ExM commented on BuildRevisionCounter/Data/MongoDBUserStorage.cs:
```diff
@@ -0,0 +1,73 @@
+		/// <returns></returns>
+		/// <exception cref="DuplicateKeyException">В случае дублирования имени пользователя</exception>
+		public async Task CreateUser(string name, string password, string[] roles)
+		{
```
Это метод тоже может возвращать bool, что упростит его использование

sFedyashov replied:
Когда правил метод создания ревизии, то думал об этом.
Что не понравилось в отказе от исключения: слишком неинформативный получается вызов - он же не TryCreateUser. В приватных методах понятно - мы сами его делаем [Try]Create и сами делаем обработку кода возврата. А в публичных методах такое проглатывание ошибки будет приводить к тому, что человек не знакомый с внутренним устройством метода, вообще не обратит на то, что метод не void и не узнает о том, что пользователь (по его параметрам) не создан (а это другие пароль и роли).

ExM replied:
- Тут не код возврата, а просто булевое поле.
- Метод можно переименовать в `TryCreateUser`
- Сигнатура метода известна, имя `TryCreateUser` достаточный признак чтобы обратить внимание на результат.
- результат работы метода:
  *\* без ошибок - реакция не требуется
  *\* DuplicateKeyException - тогда его надо явно перехватить, что ведет к увеличению вложенности кода
  *\* другие ошибки - перехватывются выше по стеку.

такой код выглядит проще

```
if(!await TryCreateUser(...))
{
// error
}
```

такой сложнее

```
try
{
    await TryCreateUser(...)
}
catch(DuplicateKeyException) //надо еще вспомнить какое именно исключение перехватывать
{
// error
}
```

ExM commented on BuildRevisionCounter.Tests/DBStorageTest.cs:
```diff
@@ -63,23 +62,24 @@ public void SetUp()
 			}
-			catch (MongoWriteException ex)
+			catch (DuplicateKeyException)
 			{
```
в NUnit есть более удобная конструкция ожидания исключения определенного типа.

sFedyashov replied:
так было изначально:

``` c#
catch (MongoWriteException ex)
{
    Assert.AreEqual(ServerErrorCategory.DuplicateKey, ex.WriteError.Category);
}
```

а так написано в требованиях:

> Codestyle священен. Код в проекте может крив, но код в запросе обязан ему соответсвовать.
> ;-)

ExM replied:
в NUnit можно пользоваться ExpectedExceptionAttribute или Asserts.Throws для более наглядного и простого способа декларации желаемого поведения. 

Я не понял ваш ответ.

ExM commented on BuildRevisionCounter.Tests/DBStorageFactory.cs:
```diff
@@ -0,0 +1,40 @@
+			var connectionString = ConfigurationManager.ConnectionStrings[connectionStringName].ConnectionString;
+			connectionString = DBUtil.SetDatabaseNameRandom(typeof (T), connectionString);
+			return (IUserDatabaseTestProvider)Activator.CreateInstance(typeof(T), connectionString);
+		}
```
Все таки, избавимся от вызовов Activator.
Генерить случайное имя БД, не следует.
В случае сбоя тестов, будут накапливаться БД, которые надо будет удалять в ручную.

sFedyashov replied:
добавлю абстрактный класс, чтобы получить требуемый конструктор.
Как раз ситуация с вопросом на собеседовании (и мой ответ про конструктор): интерфейсом тут конструктор не знаю как прикрутить.

ExM replied:
Здесь ничего не надо абстрагироватью Просто `new MongoDBUserStorage(connectionString);` достаточно.
